src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace React\HttpClient;
+
+use Evenement\EventEmitterTrait;
+use React\Stream\DuplexStreamInterface;
+use React\Stream\Util;
+
+class ChunkedStreamDecoder
+{
+    const CRLF = "\r\n";
+
+    use EventEmitterTrait;
+
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var string
+     */
+    protected $chunkExtension = '';
+
+    /**
+     * @var int
+     */
+    protected $remainingLength = 0;
+
+    /**
+     * @var bool
+     */
+    protected $nextChunkIsLength = true;
+
+    /**
+     * @var DuplexStreamInterface
+     */
+    protected $stream;
+
+    /**
+     * @param DuplexStreamInterface $stream
+     */
+    public function __construct(DuplexStreamInterface $stream)
+    {
+        $this->stream = $stream;
+        $this->stream->on('data', array($this, 'handleData'));
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+            'end',
+        ]);
+    }
+
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $this->iterateBuffer();
+        } while (strlen($this->buffer) > 0 && strpos($this->buffer, static::CRLF) !== false);
+    }
+
+    protected function iterateBuffer()
+    {
+        if ($this->nextChunkIsLength) {
+            $this->nextChunkIsLength = false;
+            $this->chunkExtension = '';
+            $lengthChunk = substr($this->buffer, 0, strpos($this->buffer, static::CRLF));
+            if (strpos($lengthChunk, ';') !== false) {
+                list($lengthChunk, $chunkExtension) = explode(';', $lengthChunk, 2);
+                $this->chunkExtension = trim($chunkExtension);
+            }
+            $this->remainingLength = hexdec($lengthChunk);
+            $this->buffer = substr($this->buffer, strpos($this->buffer, static::CRLF) + 2);
+        }
+
+        if ($this->remainingLength > 0) {
+            $chunkLength = $this->getChunkLength();
+            if ($chunkLength === 0) {
+                return;
+            }
+            $this->emit('data', array(
+                substr($this->buffer, 0, $chunkLength),
+                $this,
+                [
+                    'chunkExtension' => $this->chunkExtension,
+                ]
+            ));
```

Comment:
I'm not sold on the idea we should emit the chunk extension as part of the data message.
Do we have any specific use case for this right now?

Personally, I would rather remove this for now (reduces complexity). We can still think about introducing this once we see an actual need and we can figure out a decent API without introducing a BC break.

Also, in my projects I only emit the actual data as the other event attributes have always been unreliable across the React ecosystem and it's very easy to use a closure to add arbitrary references to the event handler anyway (e.g. https://github.com/clue/php-utf8-react/blob/master/src/Sequencer.php#L111)